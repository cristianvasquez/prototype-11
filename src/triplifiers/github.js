import rdf from 'rdf-ext'
import { matchPatterns } from './miniNLP.js'

/**
 ## URL patterns

 ```markdown
 has Github: <https://github.com/rdfjs-base/serializer-jsonld>
 ```

 ```turtle
 <https://github.com/rdfjs-base/serializer-jsonld> a :GithubRepo ;
 :label "serializer-jsonld" .
 <https://github.com/rdfjs-base> a :GithubUser ;
 :contains <https://github.com/rdfjs-base/serializer-jsonld> .
 ```
 **/


// const ns = {
//   schema: namespace('http://schema.org/'),
//   vault: namespace('http://vault.org/'),
//   rdf: namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#'),
// }

// <url> -> url
const trim = (txt) => txt.split(' ')
  .map((chunk) => chunk.replace(/^\<+|\>+$/gm, ''))
  .join(' ')

function triplify (text, baseUri, ns) {

  const results = matchPatterns(trim(text), ['has github repo? [#Url]'])
  if (results) {
    const githubUrl = results[0]
    const repoName = githubUrl.split('/').at(-1)
    const githubUserUrl = githubUrl.split('/').slice(0, -1).join('/')
    const userName = githubUrl.split('/').at(-1)

    const dataset = rdf.dataset()

    dataset.addAll([
      rdf.quad(rdf.namedNode(baseUri), ns.vault.about, rdf.namedNode(githubUrl)),
      rdf.quad(rdf.namedNode(githubUserUrl), ns.vault.contains, rdf.namedNode(githubUrl)),
      rdf.quad(rdf.namedNode(githubUserUrl), ns.rdf.type, ns.vault.GithubUser),
      rdf.quad(rdf.namedNode(githubUserUrl), ns.vault.label, rdf.literal(userName)),
      rdf.quad(rdf.namedNode(githubUrl), ns.rdf.type, ns.vault.GithubRepo),
      rdf.quad(rdf.namedNode(githubUrl), ns.vault.label, rdf.literal(repoName)),
    ])
    return dataset
  }
  return null
}

export { triplify }