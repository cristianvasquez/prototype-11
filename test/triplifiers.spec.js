import { triplify } from '../src/triplifiers/github.js'
import rdf from 'rdf-ext'
import expect from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import assert from 'assert'
expect.extend({ toMatchSnapshot })

const namespace = rdf.namespace
const baseUri = 'http://example/1'

const ns = {
  schema: namespace('http://schema.org/'),
  vault: namespace('http://vault.org/'),
  rdf: namespace('http://www.w3.org/1999/02/22-rdf-syntax-ns#'),
}

describe('Github', function () {
  describe('has Github repo', function () {
    it('has Github repo', function () {

      const content = `
   has Github repo <https://github.com/cristianvasquez/prototype-11>
`
      assert.notEqual(triplify(content, baseUri, ns), null, 'should return a dataset')
      expect(triplify(content, baseUri, ns).toString()).toMatchSnapshot(this)
    })

    it('has Github', function () {

      const content = `
   has Github <https://github.com/cristianvasquez/prototype-11>
`
      assert.notEqual(triplify(content, baseUri, ns), null, 'should return a dataset')
      expect(triplify(content, baseUri, ns).toString()).toMatchSnapshot(this)
    })

  })
})
