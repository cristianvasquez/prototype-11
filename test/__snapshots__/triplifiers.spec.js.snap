// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Github has Github repo has Github 1`] = `
"<http://example/1> <http://vault.org/about> <https://github.com/cristianvasquez/prototype-11> .
<https://github.com/cristianvasquez/prototype-11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://vault.org/GithubRepo> .
<https://github.com/cristianvasquez/prototype-11> <http://vault.org/label> \\"prototype-11\\" .
<https://github.com/cristianvasquez> <http://vault.org/contains> <https://github.com/cristianvasquez/prototype-11> .
<https://github.com/cristianvasquez> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://vault.org/GithubUser> .
<https://github.com/cristianvasquez> <http://vault.org/label> \\"prototype-11\\" .
"
`;

exports[`Github has Github repo has Github repo 1`] = `
"<http://example/1> <http://vault.org/about> <https://github.com/cristianvasquez/prototype-11> .
<https://github.com/cristianvasquez/prototype-11> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://vault.org/GithubRepo> .
<https://github.com/cristianvasquez/prototype-11> <http://vault.org/label> \\"prototype-11\\" .
<https://github.com/cristianvasquez> <http://vault.org/contains> <https://github.com/cristianvasquez/prototype-11> .
<https://github.com/cristianvasquez> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://vault.org/GithubUser> .
<https://github.com/cristianvasquez> <http://vault.org/label> \\"prototype-11\\" .
"
`;
